mmaction/models/backbones/mvit.py

```diff
@@ -0,0 +1,864 @@
+
+    Args:
+        in_channels (int): The number of input channels.
+        hidden_channels (int, optional): The number of hidden layer channels.
+            If None, same as the ``in_channels``. Defaults to None.
+        out_channels (int, optional): The number of output channels. If None,
+            same as the ``in_channels``. Defaults to None.
+        act_cfg (dict): The config of activation function.
+            Defaults to ``dict(type='GELU')``.
+        init_cfg (dict, optional): The config of weight initialization.
+            Defaults to None.
+    """
+
+    def __init__(self,
+                 in_channels,
+                 hidden_channels=None,
+                 out_channels=None,
+                 act_cfg=dict(type='GELU'),
+                 init_cfg=None):
+        super().__init__(init_cfg=init_cfg)
+        out_channels = out_channels or in_channels
+        hidden_channels = hidden_channels or in_channels
+        self.fc1 = nn.Linear(in_channels, hidden_channels)
+        self.act = build_activation_layer(act_cfg)
+        self.fc2 = nn.Linear(hidden_channels, out_channels)
+
+    def forward(self, x):
+        x = self.fc1(x)
+        x = self.act(x)
+        x = self.fc2(x)
+        return x
+
+
+def attention_pool(x: torch.Tensor,
+                   pool: nn.Module,
+                   in_size: tuple,
+                   with_cls_token: bool = False,
+                   norm: Optional[nn.Module] = None):
+    """Pooling the feature tokens.
+
+    Args:
+        x (torch.Tensor): The input tensor, should be with shape
+            ``(B, num_heads, L, C)`` or ``(B, L, C)``.
+        pool (nn.Module): The pooling module.
+        in_size (Tuple[int]): The shape of the input feature map.
+        with_cls_token (bool): Whether concatenating class token into video
+            tokens as transformer input. Defaults to True.
+        norm (nn.Module, optional): The normalization module.
+            Defaults to None.
+    """
+    ndim = x.ndim
+    if ndim == 4:
+        B, num_heads, L, C = x.shape
+    elif ndim == 3:
+        num_heads = 1
+        B, L, C = x.shape
+        x = x.unsqueeze(1)
+    else:
+        raise RuntimeError(f'Unsupported input dimension {x.shape}')
+
+    T, H, W = in_size
+    assert L == T * H * W + with_cls_token
+
+    if with_cls_token:
+        cls_tok, x = x[:, :, :1, :], x[:, :, 1:, :]
+
+    # (B, num_heads, T*H*W, C) -> (B*num_heads, C, T, H, W)
+    x = x.reshape(B * num_heads, T, H, W, C).permute(0, 4, 1, 2,
+                                                     3).contiguous()
+    x = pool(x)
+    out_size = x.shape[2:]
+
+    # (B*num_heads, C, T', H', W') -> (B, num_heads, T'*H'*W', C)
+    x = x.reshape(B, num_heads, C, -1).transpose(2, 3)
+
+    if with_cls_token:
+        x = torch.cat((cls_tok, x), dim=2)
+
+    if norm is not None:
+        x = norm(x)
+
+    if ndim == 3:
+        x = x.squeeze(1)
+
+    return x, out_size
+
+
+class MultiScaleAttention(BaseModule):
```

Comment:
What about moving this module to `models/utils/` if it can also be used in other models?

Reply:
currently, this module is not so common, we can move it when needed